### ISSUE TITLE
PaymentModule->validateOrder creates new objects in global context

### ISSUE THREAD

**Issue by RomSunZ** [PrestaShop/PrestaShop#11012]
Calling $this->module->validateOrder sets customer flag is_logged to false if it was true befor calling this function.

**To Reproduce**
1. Call $this->context->customer->isLogged(true) befor and after $this->module->validateOrder inside paymet_module/controllers/front/validation.php 
2. Check results. Before you will get true and also some guest id value, after you will get false and id_guest=0, but the same customer->id 

Inside validateOrder there are:

>  $this->context->cart = new Cart((int)$id_cart);
> $this->context->customer = new Customer((int)$this->context->cart->id_customer);
>  // The tax cart is loaded before the customer so re-cache the tax calculation method
> $this->context->cart->setTaxCalculationMethod();
> $this->context->language = new Language((int)$this->context->cart->id_lang);
> $this->context->shop = ($shop ? $shop : new Shop((int)$this->context->cart->id_shop));
> ShopUrl::resetMainDomainCache();
> $id_currency = $currency_special ? (int)$currency_special : (int)$this->context->cart->id_currency;
> $this->context->currency = new Currency((int)$id_currency, null, (int)$this->context->shop->id);

These lines are reassigning global vars, but they have been assigned before. What is the reason to reassign? 
I checked only customer logged state, maybe there are some other issues with reassigning.

**Additionnal information**
PrestaShop version: 1.7.4.2

**Comment by khouloudbelguith** [comment]
Hi @RomSunZ,

What are the exact results displayed after calling: `$this->context->customer->isLogged(true)`
did you add this line code in the /project_folder/modules/paymet_module/controllers/front/validation.php before and after `$this->module->validateOrder` ?
Thanks!

**Comment by RomSunZ** [comment]
Yes, add $this->context->customer->isLogged(true) before and after $this->module->validateOrder inside validation.php. Before you will get true and also some guest id, after you will get false and id_guest=0, but the same customer->id. I cannot check values after standart redirect to order-confirmation.php?key=... as I have not figured out where to look yet, but I think values there are OK again as you are logged in on order confirmation page.

**Comment by khouloudbelguith** [comment]
@RomSunZ, could you please add screenshot about your issue.
It will help us if you can attach a screen record.
Thanks!

**Comment by RomSunZ** [comment]
Screenshot will give you nothing. Please add this to your bankwire or cash on delivery or to what ever else payment module validate.php postProcess function and try to pay under logged customer:

> $c1=$this->context->customer;
> $l1=(int)$this->context->customer->isLogged(true);
> $this->module->validateOrder((int)$this->context->cart->id, Configuration::get('PS_OS_PREPARATION'), $total, $this->module->l('Cash on delivery', 'validation'), null, array(), null, false, $customer->secure_key);
> file_put_contents("ip.txt", 'compare customer befor and after='.(int)($this->context->customer==$c1)."\r\n", FILE_APPEND); 
> file_put_contents("ip.txt", 'customer before: id='.$c1->id.' $id_guest='.$c1->id_guest.' logged='.$l1."\r\n", FILE_APPEND); 
> file_put_contents("ip.txt", 'customer after: id='.$this->context->customer->id.' $id_guest='.$this->context->customer->id_guest.' logged='.(int)$this->context->customer->isLogged(true)."\r\n", FILE_APPEND); 
>         Tools::redirectLink(__PS_BASE_URI__.'order-confirmation.php?key='.$customer->secure_key.'&id_cart='.(int)$this->context->cart->id.'&id_module='.(int)$this->module->id.'&id_order='.(int)$this->module->currentOrder);
> 
Please check mentioned in my first post called function validateOrder. There you create new instance of Customer. There is nothing conserning customer login  state after creating new customer instance. Sorry I cannot find how to attach file, so here is copy past from ip.txt:

> compare customer befor and after=0
customer before: id=1 $id_guest=180 logged=1
customer after: id=1 $id_guest= logged=0

**Comment by khouloudbelguith** [comment]
Hi @RomSunZ,

Thanks for this clarifications.
I manage to reproduce this issue with 1.7.4.3.
after adding the line code into `validate.php`
1- Before calling: `$this->context->customer->isLogged(true)`:
I have this:
```
compare customer befor and after=0
customer before: id=10 $id_guest=18 logged=1
customer after: id=10 $id_guest= logged=0
```
2- After calling: `$this->context->customer->isLogged(true)`:
I have this:
```
compare customer befor and after=1
customer before: id=11 $id_guest= logged=0
customer after: id=11 $id_guest= logged=0
```
Thanks!

**Comment by marionf** [comment]
@khouloudbelguith Could you try to reproduce it with 1.7.6.0 ?

**Comment by khouloudbelguith** [comment]
Hi @marionf,

I have the same issue with PS1.7.6.0beta1

Before calling:` $this->context->customer->isLogged(true)`,

I have this
```
compare customer befor and after=0
customer before: id=4 $id_guest=30 logged=1
customer after: id=4 $id_guest= logged=0
```
After calling: `$this->context->customer->isLogged(true)`,

I have this

```
compare customer befor and after=0
customer before: id=5 $id_guest=31 logged=0
customer after: id=5 $id_guest= logged=0
```
Thanks!

**Comment by Hlavtox** [comment]
Fixed
